src/frontend/src/binder/expr/mod.rs

```diff
@@ -363,3 +358,29 @@ pub fn bind_data_type(data_type: &AstDataType) -> Result<DataType> {
     };
     Ok(data_type)
 }
+
+/// Find the `least_restrictive` type over a list of `exprs`, and add implicit cast when necessary.
+/// Used by `VALUES`, `CASE`, `UNION`, etc. See [PG](https://www.postgresql.org/docs/current/typeconv-union-case.html).
+pub fn align_types<'a>(exprs: impl Iterator<Item = &'a mut ExprImpl>) -> Result<DataType> {
+    use std::mem::swap;
+
+    let exprs = exprs.collect_vec();
+    // Essentially a filter_map followed by a try_reduce, which is unstable.
+    let mut ret_type = None;
+    for e in &exprs {
+        if e.is_null() {
+            continue;
+        }
+        ret_type = match ret_type {
+            None => Some(e.return_type()),
+            Some(t) => Some(least_restrictive(t, e.return_type())?),
+        };
+    }
+    let ret_type = ret_type.unwrap_or(DataType::Varchar);
+    for e in exprs {
+        let mut dummy = ExprImpl::literal_bool(false);
+        swap(&mut dummy, e);
+        *e = dummy.cast_implicit(ret_type.clone())?;
```

Comment:
If we can implement a new `cast_implicit(&self)` or impl `Default` for `ExprImpl`, I guess this part can be more readable 🤪

Reply:
Yes. But `ExprImpl::default()` does not make sense, and `cast_implicit(&self)` is not useful other than this place. Inside `cast_implicit(&self)`, it still needs to either `clone` or `swap &mut`.

Reply:
In fact, this pattern appears many times in our codebase, which is fairly acceptable when there is a default value for this type. Is there any unsafe way to handle this?

Reply:
We can implement a try_reduce by ourselves 🤣